### ISSUE TITLE
4.x: Usability for command/controller actions

### ISSUE THREAD

**Issue by dereuromark** [cakephp/cakephp#13603]
This is a (multiple allowed):

* [x] RFC

cakephp 4.x

### What you did

```php
class SomeCommand extends Command
{
    /**
     * @param \Cake\Console\Arguments $args The command arguments.
     * @param \Cake\Console\ConsoleIo $io The console io
     *
     * @return int|null The exit code or null for success
     */
    public function execute(Arguments $args, ConsoleIo $io): ?int
    {
        ...
        $io->out('Done.');
    }
```
As we used to do in 3.x - as omitting return type is the same as explicitly returning null usually.

### What happened
Of course:

> Error: [TypeError] Return value of App\Command\SomeCommand::execute() must be of the type int or null, none returned

The following (return null or int) is actually required at the end:
```php
class SomeCommand extends Command
{
    /**
     * @param \Cake\Console\Arguments $args The command arguments.
     * @param \Cake\Console\ConsoleIo $io The console io
     *
     * @return int|null The exit code or null for success
     */
    public function execute(Arguments $args, ConsoleIo $io): ?int
    {
        ...
        $io->out('Done.');

        return null;
    }
```
Same would go for controller actions and other such "action" methods.

### What you expected to happen

If we force now the return line to be present by return type hint addition, why then allowing null?
Then we can also directly go to int (static::CODE_SUCCESS is more readable then anyway).

```php
class SomeCommand extends Command
{
    /**
     * @param \Cake\Console\Arguments $args The command arguments.
     * @param \Cake\Console\ConsoleIo $io The console io
     *
     * @return int The exit code
     */
    public function execute(Arguments $args, ConsoleIo $io): int
    {
        ...
        $io->out('Done.');

        return static::CODE_SUCCESS;
    }
```

So basically my question:
Do we want to remove the typehints here to avoid having to add these return statements everywhere (commands as controllers), or do we make them non-nullable then if they cant be omitted now anymore anyway?

**Comment by markstory** [comment]
>  Do we want to remove the typehints here to avoid having to add these return statements everywhere (commands as controllers), or do we make them non-nullable then if they cant be omitted now anymore anyway?

Removing the return type entirely will be simpler for application users as they don't have to do additional work to upgrade and we can do runtime checks in `CommandRunner` to ensure we have one of the acceptable types. I agree that the current nullable int return value isn't ideal as we get all of the pain but not much of the benefit from a return type here.

**Comment by ADmad** [comment]
I too am fine with removing return typehint of `Command::execute()`. This is one case where we don't gain much by having the typehint and just causes usability issue.

**Comment by dereuromark** [comment]
Technically the typehint is then probably

    @return int|null|void The exit code or null for success

As this means it is possible to allow a void-null shortcut.

**Comment by ADmad** [comment]
Yeah we can use whatever types are required in docblock annotation to keep static analyzers happy :slightly_smiling_face:.

**Comment by markstory** [comment]
Controller actions already don't require a return type. I've put up a pull to remove the required return type on Command::execute().
